src/lib/render.c

```diff
@@ -1130,19 +1134,44 @@ notcurses_rasterize_inner(notcurses* nc, ncpile* p, FILE* out){
   if(fflush(out)){
     return -1;
   }
+#define MIN_ASU_SIZE 4096 // FIXME
+  if(*asu){
+    if(nc->rstate.mstrsize >= MIN_ASU_SIZE){
+      const char* endasu = get_escape(&nc->tcache, ESCAPE_ESU);
+      if(endasu){
+        if(fprintf(out, endasu) < 0 || fflush(out)){
+          return -1;
+        }
+      }else{
+        *asu = 0;
+      }
+    }else{
+      *asu = 0;
+    }
+  }
+#undef MIN_ASU_SIZE
   return nc->rstate.mstrsize;
 }
 
 // rasterize the rendered frame, and blockingly write it out to the terminal.
 static int
 raster_and_write(notcurses* nc, ncpile* p, FILE* out){
-  if(notcurses_rasterize_inner(nc, p, out) < 0){
+  // will we be using application-synchronized updates? if this comes back as
+  // non-zero, we are, and must emit the header. no ASU without a tty.
+  unsigned useasu = nc->ttyfd >= 0 ? true : false;
```

Comment:
no need to check tty if we check for the escape first, as we can only get it from a tty. do that.